chore: Upgrade to `typescript@5.8.3`

kitten opened the pull request:
PR that rebuilds all packages is split off for a complete CI run and reviewability: https://github.com/expo/expo/pull/36019

- Upgrade to `typescript@5.8.3`
- Fix `@expo/config/src/Config.ts`
  - Invalid `null` return value
- Fix `@expo/metro-config/src/transform-worker/css.ts`
  - New regex validation fails on escaped octals which aren't using `\x00` patterns
- Fix `package.json` and `tsconfig.json` settings for `create-expo` and `create-expo-nightly`
  - This uses `ncc` so `bundler` is the appropriate `moduleResolution` setting
- Add missing `bytes()` method to `expo/src/winter/fetch/FetchResponse`
- Add missing `ArrayBuffer` cast (to exclude `SharedArrayBuffer`) to `expo/src/devtools/DevToolsPluginClient`
- Remove redundant nullish coalescing from `expo-modules-test-core/src/mockgen.ts`
- Fix invalid context type in `expo-router/src/layouts/RootModal.tsx`
- Update `expo-router/src/rsc/router/host.tsx` to use arbitrary `Response` shaped type (since this uses `FetchResponse` but expects `Response`)
- Upgrade `typescript-eslint` packages
- Fix callback type in `@expo/cli/src/export/persistMetroAssets.ts`'s `copy` function
- Remove now unused `ts-expect-error` in `@expo/cli/src/utils/exit.ts`
- Upgrade and align to `@types/node@^22`
  - Note: I'm seeing type errors with `@types/node@^20` (which is what we'd optimally want to use), due to `SharedArrayBuffer` type incompatibilities that haven't been fixed in the upstream typings
  - For now, upgrading to `22` is stable and makes all type errors resolvable cleanly

kitten committed:
Upgrade to typescript@5.8.3

kitten committed:
Fix @expo/config/src/Config.ts' reduceExpoObject typing

kitten committed:
Reformat @expo/metro-config/src/transform-worker/css.ts' regex for octal codes

kitten committed:
Fix module settings in create-expo/tsconfig.json

kitten committed:
Fix package.json import and tsconfig.json for create-expo-nightly

kitten committed:
Add missing bytes() method to expo/src/winter/fetch/FetchResponse.ts

kitten committed:
Add missing ArrayBuffer type cast (for SharedArrayBuffer exclusion) to expo/src/devtools/DevToolsPluginClient

kitten committed:
Remove redundant nullish coalescing op in expo-modules-test-core/src/mockgen.ts

kitten committed:
Fix invalid context type in expo-router/src/layouts/RootModal.tsx

kitten committed:
Update expo-router/src/rsc/router/host.tsx to be resilient to differing Response types

kitten committed:
Upgrade typescript-eslint packages

kitten committed:
Align all @types/node versions to @types/node@^20.14.8

kitten committed:
Fix @types/node type issue in @expo/cli/src/export/persistMetroAssets.ts' copy

kitten committed:
Remove unused ts-expect-error due to `@types/node` upgrade

kitten committed:
Upgrade to `@types/node@^22` and align versions

kitten committed:
@expo/config-types: Replace generate.ts with generate.js

kitten committed:
Add missing root ts-node dependency

kitten committed:
Fix-up bash script shebang

kitten committed:
Fix/Revert tsconfig.json

kitten committed:
Fix invalid SplitConfigs default type (properties are optional)

byCedric commented on packages/expo-dev-client/plugin/tsconfig.tsbuildinfo:
Wait, is this supposed to be committed, or should this be ignored?

kitten replied:
ah, nice catch :laughing:

kitten committed:
Ignore tsconfig.tsbuildinfo

kitten committed:
Remove unused @expo/server node:http declaration

kitten committed:
Explicitly cast ArrayBuffer in src/devtools/__tests__/MessageFramePack

kitten committed:
Re-order ExpoFormData overloads

kitten committed:
Update more snapshots

kitten committed:
Update RSC flight snapshot

kitten committed:
Update html-elements snapshots

kitten committed:
Rebuild expo

byCedric commented on packages/@expo/cli/src/utils/exit.ts:
```diff
@@ -86,7 +86,6 @@ export function ensureProcessExitsAfterDelay(waitUntilExitMs = 10000, startedAtM
   ];
   // Check active resources, besides the TTYWrap/PipeWrap (process.stdin, process.stdout, process.stderr)
-  // @ts-expect-error Added in v17.3.0, v16.14.0 but unavailable in v18 typings
   const activeResources = process.getActiveResourcesInfo() as string[];
```
Thanks past-cedric to document this

byCedric commented on packages/@expo/config-types/scripts/generate.js:
```diff
@@ -1,28 +1,33 @@
-const packageJSON = require('../package.json');
+let version = '';
+
+const packageJSON = JSON.parse(fs.readFileSync(path.join(__dirname, '../package.json'), 'utf8'));
```
What's the benefit of manually reading a JSON file vs importing it?

kitten replied:
Just not having to deal with `tsconfig.json` changes. Technically, `scripts` isn't included in the `tsconfig.json`'s includes since it alters source output. Other settings could be adjusted for that, and we would also have to add `resolveJsonModules`. But to keep changes to the `tsconfig.json` to a minimum, to avoid risking some accidental breakage I'd overlook, I'm just doing a read file instead

kitten committed:
[CI/Rebuild] chore: Upgrade to `typescript@5.8.3` (#36019)

Rebuilding all packages to let CI run through the changes in the base
branch

This only contains an `et check-packages` build run and manual changes
are in: https://github.com/expo/expo/pull/36016
